### ISSUE TITLE
Document's embeddings disappear after the structs array update

### ISSUE THREAD

**Issue by JankowskiDaniel** [vespa-engine/vespa#29675]
**Describe the bug**
Having defined schema for multi-vector search with an additional array of structs field, after adding a new struct into the array, the embeddings disappear for this record. 

**To Reproduce**
1. A part of defined schema:
```schema bert_search {
    document bert_search {
        struct topic {
            field nametype string {}
            field surname type string {}

      field chunks_embeddings type tensor<float>(p{}, x[384]) {
                  indexing: attribute | index | summary
                  attribute {
                      distance-metric: angular
                  }
                  index {
                      hnsw {
                    max-links-per-node: 16
                    neighbors-to-explore-at-insert: 200
                }
            }
       ... <other declared fields>
        }
        field topics type array<topic> {
            indexing: summary
            struct-field name {
                indexing: attribute
                attribute: fast-search
            }
            struct-field surname{
                indexing: attribute
                attribute: fast-search
            }
```
2. Take any indexed document, e.g. `select * from bert_search where true`. The sample response is as follows:
```
"children": [
            {
                "id": "id:bert_search:bert_search::1",
                "relevance": 0.0,
                "source": "dcsearch_content",
                "fields": {
                    "sddocname": "bert_search",
                    "documentid": "id:bert_search:bert_search::1",
                    "hash": "1234",
                    "topics": [
                       {
                           "name": "John",
                           "surname": "Smith"
                       }]
                    "chunks_embeddings": {
                        "type": "tensor<float>(p{},x[384])",
                        "blocks": {
                            "1": [ ... ]
```
3. Update the document with the following request:
```
curl --location --request PUT 'http://localhost:8080/document/v1/bert_search/bert_search/docid/1' \
     --header 'Content-Type: application/json' \
     --data '{
    "update": "id:bert_search:bert_search::1",
    "fields":
        {
            "topics": {
                "add": [
                        {
                            "name": "John",
                            "surname": "Williams"
                        }
                    ]
            }
        }
}
```
4. Check the document content after the operation with a query `select * from bert_search where hash contains "1234"`. Although the `topics` array has been updated correctly, the `chunks_embeddings` field disappeared from the record.
```
"children": [
            {
                "id": "id:bert_search:bert_search::1",
                "relevance": 0.0,
                "source": "dcsearch_content",
                "fields": {
                    "sddocname": "bert_search",
                    "documentid": "id:bert_search:bert_search::1",
                    "hash": "1234",
                    "topics": [
                       {
                           "name": "John",
                           "surname": "Smith"
                       },
                       {
                           "name": "John",
                           "surname": "Williams"
                       }]
                 }
              }
            ]
```
**Expected behavior**
The `chunks_embeddings` field should be present for the records after performing a partial update operation

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Infrastructure: Vespa application deployed with Docker
 - Versions: Ubuntu: 20.04.6 LTS

**Vespa version**
Vespa CLI version 8.256.22 compiled with go1.19.13 on linux/amd64

**Additional context**
The important fact in this context might be that the embeddings are added to the record after the record itself is created, i.e. in this use case a record, first is indexed for the BM25 mode, then it is updated with embeddings to make it searchable in the semantic mode, at the end the record is modified as described above.

**Comment by jobergum** [comment]
I'm not able to reproduce this. The instructions above contained probably obfuscated field names so I based the reproduction effort on https://github.com/vespa-engine/sample-apps/tree/master/multi-vector-indexing 

Added a struct 

```
struct topic {
        field nametype type string {}
        field name  type string {}
	}
    field topics type array<topic> {
        indexing: summary
        struct-field nametype {
            indexing: attribute
            attribute: fast-search
        }
        struct-field name {
            indexing: attribute
            attribute: fast-search
        }
    }
```

And partial updates works against topics works as expected. 

```zstdcat ext/articles.jsonl.zst |vespa feed -```

Run a query 

```
vespa query 'yql=select * from wiki where userQuery() or ({targetHits:1}nearestNeighbor(paragraph_embeddings,q))' \
  'input.query(q)=embed(what does 24 mean in the context of railways)' \
  'query=what does 24 mean in the context of railways' \
  'ranking=hybrid' \
  'hits=1'
```

Gets the output

```
vespa query 'yql=select id from wiki where userQuery() or ({targetHits:1}nearestNeighbor(paragraph_embeddings,q))' \
  'input.query(q)=embed(what does 24 mean in the context of railways)' \
  'query=what does 24 mean in the context of railways' \
  'ranking=hybrid' \
  'hits=1' 
{
    "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 8
        },
        "coverage": {
            "coverage": 100,
            "documents": 8,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        },
        "children": [
            {
                "id": "index:articles/0/d69bc0b11402253b646d9802",
                "relevance": 4.163399168193791,
                "source": "articles",
                "fields": {
                    "matchfeatures": {
                        "bm25(paragraphs)": 10.468827250036052,
                        "bm25(title)": 1.1272217840066165,
                        "closest(paragraph_embeddings)": {
                            "4": 1.0
                        },
                        "firstPhase": 0.8807156260391702,
                        "all_paragraph_similarities": {
                            "1": 0.8030083179473877,
                            "2": 0.7992785573005676,
                            "3": 0.8273358345031738,
                            "4": 0.8807156085968018,
                            "0": 0.849757194519043
                        },
                        "avg_paragraph_similarity": 0.8320191025733947,
                        "max_paragraph_similarity": 0.8807156085968018
                    }
                }
            }
        ]
    }
}
```

Then run the update operation, adding a element to the array struct
```
cat update.json 
{
    "update": "id:wikipedia:wiki::9985",
    "fields": {
        "topics": {
            "add": [
                {
                    "name": "foo",
                    "nametype": "bar"
                }
            ]
        }
    }
}
```
And run a query again, now also returning the array of structs and the vector similarities stays unchanged. Vector data is there also.

```
vespa query 'yql=select id, topics from wiki where userQuery() or ({targetHits:1}nearestNeighbor(paragraph_embeddings,q))' \
  'input.query(q)=embed(what does 24 mean in the context of railways)' \
  'query=what does 24 mean in the context of railways' \
  'ranking=hybrid' \
  'hits=1'
{
    "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 8
        },
        "coverage": {
            "coverage": 100,
            "documents": 8,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        },
        "children": [
            {
                "id": "index:articles/0/d69bc0b11402253b646d9802",
                "relevance": 4.163399168193791,
                "source": "articles",
                "fields": {
                    "matchfeatures": {
                        "bm25(paragraphs)": 10.468827250036052,
                        "bm25(title)": 1.1272217840066165,
                        "closest(paragraph_embeddings)": {
                            "4": 1.0
                        },
                        "firstPhase": 0.8807156260391702,
                        "all_paragraph_similarities": {
                            "1": 0.8030083179473877,
                            "2": 0.7992785573005676,
                            "3": 0.8273358345031738,
                            "4": 0.8807156085968018,
                            "0": 0.849757194519043
                        },
                        "avg_paragraph_similarity": 0.8320191025733947,
                        "max_paragraph_similarity": 0.8807156085968018
                    },
                    "topics": [
                        {
                            "name": "foo",
                            "nametype": "bar"
                        }
                    ]
                }
            }
        ]
    }
}
```

**Comment by jobergum** [comment]
I have also verified that

1. feed the initial docs without the `paragraphs` field
2. update the doc with the paragraphs field to compute embeddings using partial update and assigning the entire array, to invoke the embedder and create the vectors. 
3. Then, perform the partial structural update. No regressions were detected. 

So, we need a better explanation of how to reproduce this.

**Comment by JankowskiDaniel** [comment]
Hi, thank you for the response. I'll provide in detail the specification how to reconstruct above issue. Let me also provide screenshots form Postman to show responses from Vespa that I had.

In the following schema, I've provide some dumb and sample fields names, however, in my use case, the structure schema is the same. 

```
schema bert_search {
    document bert_search {

        struct topic {
            field sample_field_1 type string {}
            field sample_field_2 type string {}
            field sample_field_3 type string {}
            field sample_field_4 type string {} 
            field sample_field_5 type string {}
            field sample_field_6 type string {}
            field sample_field_7 type string {}
            field sample_field_8 type string {}
            field sample_field_9 type string {} 
            field sample_field_10 type string {}
            field sample_field_11 type string {}
            field sample_field_12 type string {}
            field sample_field_13 type string {}
            field sample_field_14 type string {}
            field sample_field_15 type string {}
            field sample_field_16 type string {}
            field sample_field_17 type string {}
            field sample_field_18 type byte {}
            field sample_field_19 type string {}
            field sample_field_20 type string {}
            field sample_field_21 type string {}
            field sample_field_22 type string {}
            field sample_field_23 type string {}
            field sample_field_24 type string {}
            field sample_field_25 type string {}
            field sample_field_26 type string {}
            field sample_field_27 type string {}
            field sample_field_28 type string {}
            field sample_field_29 type string {}
            field sample_field_30 type string {}
            field sample_field_31 type string {}
            field sample_field_32 type string {}
        }

        field main_field_1 type string {
            indexing: attribute | summary
        }
        field main_field_2 type string {
            indexing: attribute | summary
        }
        field chunks type array<string> {
            indexing: summary | index
            index: enable-bm25
        }
        field main_field_3 type string {
            indexing: attribute | summary
        }
        field main_field_4 type string {
            indexing: summary
        }

        field chunks_embeddings type tensor<float>(p{}, x[384]) {
            indexing: attribute | index | summary
            attribute {
                distance-metric: angular
            }
            index {
                hnsw {
                    max-links-per-node: 16
                    neighbors-to-explore-at-insert: 200
                }
            }

        }
        field topics type array<topic> {
            indexing: summary
            struct-field sample_field_1 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_2 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_3 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_4 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_5 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_6 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_7 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_8 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_9 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_10 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_11 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_12 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_13 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_14 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_15 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_16 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_17 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_18 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_19 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_20 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_21 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_22 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_23 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_24 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_25 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_26 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_27 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_28 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_29 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_30 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_31 {
                indexing: attribute
                attribute: fast-search
            }
            struct-field sample_field_32 {
                indexing: attribute
                attribute: fast-search
            }
        }
    }


    fieldset default {
        fields: chunks
    }

    rank-profile bm25 inherits default {
        first-phase {
            expression : bm25(chunks)
        }
        match-features: bm25(chunks)
    }

    rank-profile semantic inherits default {
        inputs {
            query(query_embedding) tensor<float>(x[384])
        }
        first-phase {
            rank-score-drop-limit: 0.3
            expression: cos(distance(field, chunks_embeddings))
        }
        match-features: closest(chunks_embeddings)
    }

    rank-profile hybrid inherits default {
        inputs {
            query(query_embedding) tensor<float>(x[384])
        }
        first-phase {
            expression: bm25(chunks) + closeness(field, chunks_embeddings)
        }
        match-features {
            closest(chunks_embeddings)
            bm25(chunks)
        }
    }

}
```

The first operation is to insert a sample document, however, without  paragraph's embeddings.

```
curl --location 'localhost:1542/document/v1/bert_search/bert_search/docid/1' \
--header 'Content-Type: application/json' \
--data '{
    "fields": {
        "main_field_1": "abc",
        "main_field_2": "abc",
        "main_field_3": "abc",
        "main_field_4": "abc",
        "chunks": ["abc", "abc", "abc", "abc"],
        "topics": [
            {
                "sample_field_1": "abc",
                "sample_field_2": "abc",
                "sample_field_3": "abc",
                "sample_field_4": "abc",
                "sample_field_5": "abc",
                "sample_field_6": "abc",
                "sample_field_7": "abc",
                "sample_field_8": "abc",
                "sample_field_9": "abc",
                "sample_field_10": "abc",
                "sample_field_11": "abc",
                "sample_field_12": "abc",
                "sample_field_13": "abc",
                "sample_field_14": "abc",
                "sample_field_15": "abc",
                "sample_field_16": "abc",
                "sample_field_17": "abc",
                "sample_field_18": 1,
                "sample_field_19": "abc",
                "sample_field_20": "abc",
                "sample_field_21": "abc",
                "sample_field_22": "abc",
                "sample_field_23": "abc",
                "sample_field_24": "abc",
                "sample_field_25": "abc",
                "sample_field_26": "abc",
                "sample_field_27": "abc",
                "sample_field_28": "abc",
                "sample_field_29": "abc",
                "sample_field_30": "abc",
                "sample_field_31": "abc",
                "sample_field_32": "abc"
            }
        ]
    }
}
```

The document has been properly inserted. Now, I want to update embeddings fields for this document with pre-computed embeddings from the external transformers model. The operation is as follows:

```
curl --location --request PUT 'localhost:1542/document/v1/bert_search/bert_search/docid/1' \
--header 'Content-Type: application/json' \
--data '{
    "fields": {
        "chunks_embeddings": {
            "assign": {"0": [0.6630910886974709, 0.9486660127685047, 0.4526558152845347, 0.48743964721515076, ...], "1": [0.3350800653178214, 0.876444464525327, 0.6390878598314982, 0.7884487197818228, ...], "2": [0.003827372837377885, 0.444636197030416, 0.19438756067580876, 0.63631119818055, ...], "3": [0.37971387132236845, 0.05141873263915353, 0.988765888715887, 0.18838058250174716, ...]}
        }
    }
}
```

The result of above operation seems to working good, I have embeddings for that document, here is a result after querying the vespa:

![image](https://github.com/vespa-engine/vespa/assets/54834564/56193956-8020-49b3-8b5e-b968fc0eac8d)

Now, lets update this document by adding a new struct element with a such request:

```
curl --location --request PUT 'localhost:1542/document/v1/bert_search/bert_search/docid/1' \
--header 'Content-Type: application/json' \
--data '{
    "fields": {
        "topics": {
            "add":
            [{
                "sample_field_1": "cde",
                "sample_field_2": "cde",
                "sample_field_3": "cde",
                "sample_field_4": "cde",
                "sample_field_5": "cde",
                "sample_field_6": "abc",
                "sample_field_7": "abc",
                "sample_field_8": "abc",
                "sample_field_9": "abc",
                "sample_field_10": "abc",
                "sample_field_11": "abc",
                "sample_field_12": "abc",
                "sample_field_13": "abc",
                "sample_field_14": "abc",
                "sample_field_15": "abc",
                "sample_field_16": "abc",
                "sample_field_17": "abc",
                "sample_field_18": 1,
                "sample_field_19": "abc",
                "sample_field_20": "abc",
                "sample_field_21": "abc",
                "sample_field_22": "abc",
                "sample_field_23": "abc",
                "sample_field_24": "abc",
                "sample_field_25": "abc",
                "sample_field_26": "abc",
                "sample_field_27": "abc",
                "sample_field_28": "abc",
                "sample_field_29": "abc",
                "sample_field_30": "abc",
                "sample_field_31": "abc",
                "sample_field_32": "abc"
            }
        ]
    }
}
}
```

Now, after making a query again, there is no embeddings, that previously has been added to this document. Here is the json response from Vespa. A new struct element has been added properly, however, the embeddings are missing.

```
{
    "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 1
        },
        "coverage": {
            "coverage": 100,
            "documents": 1,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        },
        "children": [
            {
                "id": "id:bert_search:bert_search::1",
                "relevance": 0.0,
                "source": "dcsearch_content",
                "fields": {
                    "sddocname": "bert_search",
                    "documentid": "id:bert_search:bert_search::1",
                    "main_field_1": "abc",
                    "main_field_2": "abc",
                    "chunks": [
                        "abc",
                        "abc",
                        "abc",
                        "abc"
                    ],
                    "main_field_3": "abc",
                    "main_field_4": "abc",
                    "topics": [
                        {
                            "sample_field_1": "abc",
                            "sample_field_10": "abc",
                            "sample_field_11": "abc",
                            "sample_field_12": "abc",
                            "sample_field_13": "abc",
                            "sample_field_14": "abc",
                            "sample_field_15": "abc",
                            "sample_field_16": "abc",
                            "sample_field_17": "abc",
                            "sample_field_18": 1,
                            "sample_field_19": "abc",
                            "sample_field_2": "abc",
                            "sample_field_20": "abc",
                            "sample_field_21": "abc",
                            "sample_field_22": "abc",
                            "sample_field_23": "abc",
                            "sample_field_24": "abc",
                            "sample_field_25": "abc",
                            "sample_field_26": "abc",
                            "sample_field_27": "abc",
                            "sample_field_28": "abc",
                            "sample_field_29": "abc",
                            "sample_field_3": "abc",
                            "sample_field_30": "abc",
                            "sample_field_31": "abc",
                            "sample_field_32": "abc",
                            "sample_field_4": "abc",
                            "sample_field_5": "abc",
                            "sample_field_6": "abc",
                            "sample_field_7": "abc",
                            "sample_field_8": "abc",
                            "sample_field_9": "abc"
                        },
                        {
                            "sample_field_1": "cde",
                            "sample_field_10": "abc",
                            "sample_field_11": "abc",
                            "sample_field_12": "abc",
                            "sample_field_13": "abc",
                            "sample_field_14": "abc",
                            "sample_field_15": "abc",
                            "sample_field_16": "abc",
                            "sample_field_17": "abc",
                            "sample_field_18": 1,
                            "sample_field_19": "abc",
                            "sample_field_2": "cde",
                            "sample_field_20": "abc",
                            "sample_field_21": "abc",
                            "sample_field_22": "abc",
                            "sample_field_23": "abc",
                            "sample_field_24": "abc",
                            "sample_field_25": "abc",
                            "sample_field_26": "abc",
                            "sample_field_27": "abc",
                            "sample_field_28": "abc",
                            "sample_field_29": "abc",
                            "sample_field_3": "cde",
                            "sample_field_30": "abc",
                            "sample_field_31": "abc",
                            "sample_field_32": "abc",
                            "sample_field_4": "cde",
                            "sample_field_5": "cde",
                            "sample_field_6": "abc",
                            "sample_field_7": "abc",
                            "sample_field_8": "abc",
                            "sample_field_9": "abc"
                        }
                    ]
                }
            }
        ]
    }
}
```

**Comment by jobergum** [comment]
Thank you for better instructions on how to reproduce. In this case you are using `add`, and not `assign` - not sure if it makes any difference. This is not expected behaviour, so I will attempt to reproduce it locally and follow up on this in January.

Unrelated, but partial updates against struct fields in Vespa are not true partial like for regular attribute fields, and triggers read, apply, write back pattern (which likely triggers this bug). It's not very efficient and it's faster to just send the full document if that is available.

**Comment by JankowskiDaniel** [comment]
Thank you. I believe that I have checked both `add` and `assign` modes and there is no difference for this problem. For now, to omit this issue, we were taking embeddings and then reproducing them one more time. So instead of doing `add` operation, we took the old struct array, then we updated it with the new element, and at the end `assign` a modified array value together with the `chunks_embeddings` field. However, it looks non-efficient, because we have to make additional query to first select the old array state and embeddings.

**Comment by jobergum** [comment]
I'm able to reproduce the behavior with your instructions.

Any partial update that triggers the read-write-pattern [doc](https://docs.vespa.ai/en/partial-updates.html) seems to trigger the behavior, but it's not triggered if `chunks_embeddings` was in the initial add. 

For example this works as expected

- put with fields and the `chunk_embeddings` 
- update any field that triggers read-apply-write (`main_field_4` for example)

While this fails 
- put with fields without the `chunk_embeddings` 
- assign the chunk_embeddings
- update any field that triggers read-apply-write (`main_field_4` for example)

**Comment by jobergum** [comment]
In addition, this only happens when `index` is enabled for the embedding field (hnsw indexing)

``` 
field chunks_embeddings type tensor<float>(p{}, x[2]) {
            indexing: attribute | summary
            attribute {
                distance-metric: angular
            }
}
```

**Comment by arnej27959** [comment]
this is fixed in current vespa version, with system test here:

[PartialUpdate::test_disappearing_tensor](
https://github.com/vespa-engine/system-test/blob/fc28e2711c676746c1c3ba85fc58a3bc239de2da/tests/search/partialupdate/partialupdate.rb#L528)

**Comment by geirst** [comment]
Fixed in Vespa 8.291.13.
